### ISSUE TITLE
[Bug] MTP draft model does not inherit --hf-overrides (YaRN): length gate + dummy-0 drafts silently kill speculative acceptance beyond the draft's max_model_len

### ISSUE THREAD

**Issue by wenzhelee** [vllm-project/vllm#58080]
# [Bug] MTP draft model does not inherit `--hf-overrides` (YaRN): length gate + dummy-0 drafts silently kill speculative acceptance beyond the draft's `max_model_len`

## Summary

When a model's context is extended with `--hf-overrides` **as a dict** (e.g. YaRN ×4), the MTP draft model's `model_config` is **not** given the override — `compose_draft_hf_overrides()` documents dict overrides as "target-specific key patches and are not applied to the draft". The draft therefore keeps the model's native `max_position_embeddings` as its `max_model_len`.

That length feeds the drafter's input gate in `gpu_model_runner` (`_input_fits_in_drafter`): once the batch's longest sequence + `num_spec_tokens` exceeds the **draft's** `max_model_len`, vLLM **skips draft proposal entirely and substitutes dummy token 0 for every request in the batch**. Since `rejection_sampler` reserves only `-1` as placeholder, token 0 is an ordinary token and is "accepted" only when the target happens to sample 0 — ≈0%. Result: engine-wide MTP acceptance collapses to ~0 (per-position 0.000/0.000) for as long as any sequence is past the boundary, each decode step still pays the target-side cost of verifying the dummies, and the only signal is the spec-decode metrics. Target output stays correct; nothing is logged.

Note (correction to earlier text in this issue): the RoPE objects themselves are **not** what diverges — the MTP layer builds its rope from the target's config (`initialize_model` never forwards the draft `model_config` to the model class, and `qwen3_5_mtp` reads `vllm_config.model_config.hf_text_config`). What diverges is the **draft's `max_model_len`**, which drives the gate. Thanks to @jahnclawdmonet for tracing this.

## Environment

- vLLM **0.26.0** (V1 engine), image `vllm/vllm-openai:latest` (commit ffb2d59b1c05); code paths verified in-container at the line numbers below, same code reported on main
- Model: `Qwen3.8-27B-FP8` — Qwen3-Next-style **hybrid GDN** (48 linear-attention GDN layers + 16 full-attention layers + 1 MTP draft layer), fp8 weights, fp8 KV cache
- Hardware: 2× NVIDIA H20 96 GB, `--tensor-parallel-size 2`
- Key flags: `--max-model-len 1000000`, `--speculative-config '{"method":"mtp","num_speculative_tokens":2}'`, `--enable-prefix-caching`, `--enable-chunked-prefill`, `--kv-cache-dtype fp8`, `--reasoning-parser qwen3`, and:

```
--hf-overrides '{"text_config":{"rope_parameters":{"mrope_interleaved":true,"mrope_section":[11,11,10],"rope_type":"yarn","rope_theta":10000000,"partial_rotary_factor":0.25,"factor":4.0,"original_max_position_embeddings":262144}}}'
```

(i.e. YaRN ×4: native 262,144 → 1,000,000 max sequence length.)

## Mechanism (verified in v0.26.0 source)

1. `vllm/config/speculative.py`, `compose_draft_hf_overrides()`: dict overrides are not applied to the draft. Startup logs accordingly show
   - target: `Using max model len 1000000`, `rope_type=yarn, factor=4.0`
   - **MTP draft: `Using max model len 262144`** (native config)
2. The draft's rope is *not* the mismatch: `initialize_model` (`vllm/model_executor/model_loader/utils.py:63`) passes only `vllm_config` to new-style model classes — the draft `model_config` parameter is accepted but never forwarded; `qwen3_5_mtp.py:70,73` reads `vllm_config.model_config.hf_text_config` (the target's, overridden config), and `_ROPE_DICT` (`rotary_embedding/__init__.py:30,83`) hands the MTP layers the target's rope object itself.
3. The gate: `gpu_model_runner.py:665-669` sets `effective_drafter_max_model_len = draft_config.max_model_len`; `_input_fits_in_drafter` (`:4481-4493`) returns `common_attn_metadata.max_seq_len + num_spec_tokens <= effective_drafter_max_model_len` (MTP: `+2`, no DFlash bonus). `max_seq_len` is the **batch's** longest total sequence (prompt + generated, counted optimistically).
4. On gate failure: `:4601` skips `propose_draft_token_ids`; `:4647-4654` substitutes `torch.zeros(...).expand(len(req_ids), num_spec_tokens)` — dummy token 0 for **every request in the batch** (the in-code comment cites protecting Mamba recurrent state from stale drafts).
5. `rejection_sampler.py:30`: `PLACEHOLDER_TOKEN_ID = -1` is reserved — token 0 is ordinary. A dummy survives only if the target's argmax is 0 (greedy kernel) or the random kernel samples token 0 at its own probability: ≈0, without being structurally 0.
6. Batch-wide dead regime: because the gate reads the batch maximum and the dummy expansion covers all `req_ids`, **one long session holds every concurrent request in the dead regime until it finishes** — matching the observed exactly-0.000 per-position rates engine-wide, including for short requests.

### Bisection of the crossover (same engine, same request shape)

| prompt length | mean acceptance length | verdict |
|---|---|---|
| 243,000 tok | 1.74 – 3.00 | healthy |
| 279,000 tok | 1.00 (0% accepted) | dead |
| 312,000 tok | 1.00 (0.05%) | dead |

The crossover sits just below 262,142 (draft `max_model_len` − 2) and slides lower as the session generates — consistent with the table (279 K prompt + a few hundred generated already crosses).

### Signature in engine logs

While a >262 K sequence decodes, the per-10 s `SpecDecoding metrics` line shows:

```
Mean acceptance length: 1.00, Accepted throughput: 0.00 tokens/s,
Drafted throughput: ~190 tokens/s, Per-position acceptance rate: 0.000, 0.000,
Avg Draft acceptance rate: 0.0%
```

while `vllm:num_requests_running`, TTFT and KV usage look normal. When the long sequence finishes, acceptance recovers to 2.5+ on its own — which initially made this look workload/load-related.

## Impact

- On a shared engine with mixed traffic, decode throughput oscillates with the longest live session: with 2 concurrent ~312 K-token agent sessions we measured engine-wide acceptance 0.05% (mean length 1.00) and ~60–90 tok/s total, vs. ~150–170 tok/s for comparable load once the draft was fixed.
- The dead regime is **batch-wide**, not per-request: all concurrent users lose the spec-decode benefit (and pay the dummy-verification cost) while any one session is past the boundary.
- Very hard to diagnose: no error/warning anywhere; target output remains correct; the only signals are the spec-decode metrics or "it got slower after our sessions grew past X tokens".

## Workaround we used (model-file hack)

Baking the rope extension into the model's `config.json` (`text_config.rope_parameters`: `rope_type=yarn, factor=4.0, original_max_position_embeddings=262144`) so the draft's native config already carries it:

- draft startup log: `Using max model len 262144` → `1048576` (derived length 262144×4, clamped to the target's 1,000,000) → the gate stops firing
- 279 K request: mean acceptance length 1.00 → 2.37–2.72
- 2×312 K concurrent: 0.05% acceptance → 2.27–2.35, 153 tok/s
- no regression below 262 K (2.30–3.00)

This works but requires editing the model distribution — a user should not have to know this.

## Suggested fix

1. **Propagate dict `hf_overrides` to the draft's `model_config`** (at least the length-affecting fields: `max_position_embeddings` / rope scaling), so the draft's `max_model_len` tracks the target's extended length and the gate stops firing. This is exactly what the `config.json` bake achieves manually.
2. At minimum, **warn at startup** when speculative decoding is enabled and `draft_model_config.max_model_len` < the target's effective `max_model_len` (i.e. the condition the gate actually uses — not the native `max_position_embeddings`). This path currently emits no log. Analogous to the existing warning at `speculative.py:901` for `num_speculative_tokens > 1`.

## Reproduction steps (minimal)

1. Serve any MTP-capable model whose native `max_position_embeddings` is smaller than the extended length, extending context with a **dict** `--hf-overrides` (e.g. YaRN), with `--speculative-config '{"method":"mtp","num_speculative_tokens":2}'`.
2. Confirm startup logs show different `Using max model len` for target vs. MTP draft (draft = native).
3. Send a chat completion whose prompt token count puts the total sequence above the draft's `max_model_len` (crossover ≈ draft `max_model_len` − 2).
4. Observe `vllm:spec_decode_num_accepted_tokens_total` stop growing (Δ≈0 while Δdrafts≈2 per step) and `Per-position acceptance rate: 0.000, 0.000` — for **all** requests in the batch, not just the long one.
5. Control: same request below the threshold → acceptance healthy.

## Related (distinct) issues

- #54453 — quadratic decode under structured output + reasoning parser (CPU-side; fixed by #55223, which we also backported; not the cause of this one)
- #53912 — hybrid + MTP output corruption
- #55766 — hybrid GDN + mamba `align` + prefix-cache hit → NaN logits

This issue is a fourth, silent MTP failure mode specific to **target/draft `max_model_len` divergence** via dict `hf_overrides`; it affects any hybrid or dense model that extends context that way with MTP enabled.

**Comment by jahnclawdmonet** [comment]
The rope difference is not what you are hitting. The dropped `hf_overrides` are real, but the MTP layer builds its rope from the target's config: `initialize_model` hands `vllm_config` to the model class and never forwards the draft `model_config` it was given (`model_loader/utils.py:63`), so `qwen3_5_mtp.py:70,73` reads `vllm_config.model_config.hf_text_config`, the overridden one. With `_ROPE_DICT` keyed on those parameters (`rotary_embedding/__init__.py:30,83`) the MTP layer ends up on the target's rope object itself.

What the dropped override does change is the draft's `max_model_len`, the 262,144 in your startup log, and that feeds a length gate.

`gpu_model_runner.py:665-667` sets `effective_drafter_max_model_len` from `draft_model_config.max_model_len`. `_input_fits_in_drafter` at `:4481` compares the batch against it:

```python
        num_drafter_query_tokens = self.num_spec_tokens + (
            1 if self.speculative_config.use_dflash() else 0
        )
        return (
            common_attn_metadata.max_seq_len + num_drafter_query_tokens
            <= self.effective_drafter_max_model_len
        )
```

`use_dflash()` is false for `mtp`, so the test is `max_seq_len + 2 <= 262144`. Since `max_seq_len` is the batch's longest total sequence, counting prompt plus generated tokens and counting them optimistically, the crossover prompt length sits below 262,142 and drops as a session generates.

On failure `:4601` skips `propose_draft_token_ids` and `:4647` substitutes dummies:

```python
                self._draft_token_ids = torch.zeros(
                    1, device=self.device, dtype=torch.int32
                ).expand(len(self.input_batch.req_ids), self.num_spec_tokens)
```

Token id 0 is not special to the rejection sampler, which reserves `-1` for that (`rejection_sampler.py:30`), and `_draft_probs` is set to None on the next line so `NO_DRAFT_PROBS` holds. In the greedy kernel a dummy survives only if the target's argmax is 0 (`:759`); in the random kernel it is accepted with the target's own probability for token 0 (`:818-831`). Near zero either way, without being structurally zero. Your 279 K row reads 0% and your 312 K row 0.05%.

The `.expand` covers every request in the batch and the gate reads the batch maximum. One long session holds every concurrent request in the dead regime until it finishes. In that regime no draft proposal runs at all, and the remaining cost is the target verifying the dummies.

Both of your suggested fixes hold up. Propagating the rope overrides works for the same reason your `config.json` bake did: the draft's derived length becomes 1,048,576, clamped to the target's 1,000,000, and the gate stops firing. The startup warning is worth having because this path emits no log. Its proposed condition checks the draft's native `max_position_embeddings`; the gate checks `draft_model_config.max_model_len`.

Line numbers are v0.26.0; the same code is on main.

**Comment by wenzhelee** [comment]
@jahnclawdmonet Thanks — this is exactly right, and I verified every cited line in our v0.26.0 container before rewriting the issue:

- `model_executor/model_loader/utils.py:63` — `initialize_model` passes only `vllm_config` to new-style model classes; the draft `model_config` param is never forwarded ✓
- `qwen3_5_mtp.py:70,73` — reads `vllm_config.model_config.hf_text_config` (target's overridden config) ✓
- `gpu_model_runner.py:665-669` (`effective_drafter_max_model_len = draft_config.max_model_len`), `:4481-4493` (gate), `:4601` (skip), `:4647-4654` (dummy zeros + the Mamba comment) ✓
- `rejection_sampler.py:30` — `PLACEHOLDER_TOKEN_ID = -1` ✓

So the issue body's earlier "rope divergence" mechanism is corrected to the gate + dummy-0 path, with your code references. Two of your points explain things our data showed but we couldn't account for:

1. **Batch-wide dead regime**: we saw per-position 0.000/0.000 engine-wide (including for short requests sharing the batch) while a 307 K session decoded — gate reads batch `max_seq_len`, dummy `.expand` covers all `req_ids`.
2. **"Near zero, not structurally zero"**: our 312 K row read 0.05% and we saw occasional single acceptances in 10 s windows (target genuinely sampling token 0) — matches the random-kernel path.

Our `config.json` bake was in fact equivalent to the propagation fix you describe: the draft's derived length became 1,048,576, the gate stopped firing, and 2×312 K concurrent went from 0.05% acceptance / mean length 1.00 to 2.27–2.35.

Agreed on the warning condition correction: it should compare `draft_model_config.max_model_len` against the target's effective `max_model_len` (what the gate uses), not the native `max_position_embeddings`.

One question for the maintainers: the dummy-0 fallback comment cites protecting Mamba recurrent state from stale drafts (Nemotron-H). If the gate is a safety mechanism for near-max-length inputs, is propagating dict `hf_overrides` into the draft's `model_config` (at least the length-affecting fields) the intended fix, or is the draft `max_model_len` deliberately clamped to the native config? Happy to test a patch either way.
